proposals/route-blinding.md

```diff
@@ -0,0 +1,242 @@
+  E(i+1) = H(E(i) || ss(i)) * E(i)
+  Forward the onion to the next node and include E(i+1) in a TLV field in the message extension
+```
+
+### Receiving from a blinded route
+
+When `N(r)` receives the onion message and `E(r)` in the TLV extension, she does the same
+unwrapping as intermediate nodes. The difference is that the onion will be a final onion.
+
+### Unblinding channels via fee probing
+
+The fees and cltv for the blinded route can be abused by the sender to try to unblind the real
+nodes and channels used. The sender can create onions with increased fees/cltv for the first
+blinded hop, starting with very low values. While the fee/cltv is below the real fee of the first
+hop, the sender will get an error from `N(0)`. Once the fee/cltv proposed actually satisfies the
+first hop's requirements, the error will come from another node `N(i)` inside the blinded path.
+
+The sender can then unblind channels one-by-one by discovering their real fees/cltv and matching
+those to existing channels in the graph.
+
+To mitigate this, when nodes along the blinded path are offered an invalid HTLC, they should:
+
+* return a dummy error encrypted with a throw-away key: the sender will receive an error she can't
+  decrypt and doesn't know which node generated it
+* hold the HTLC for a random amount of time before sending the error (otherwise the sender can
+  still use timing to guess what node errored out)
+
+Even with such mitigations the sender can discover the real fees/cltv of one of the blinded
+channels. To do so she uses the correct fees and cltv for all but one channel, and for that target
+channel she tries fees/cltv until the payment succeeds. Once the payment succeeds she knows the
+approximate fees/cltv of the target channel (but since the payment succeeded, she can't continue
+probing).
+
+Maybe nodes along the blinded path could use slightly different fees/cltv than what they publicly
+advertize? Or the recipient could add some fuzzing to it to blind them more?
+
+Are those mitigations enough? Or can a clever attacker still work around them?
+
+## Tips and Tricks
+
+### Dummy hops
+
+The sender knows an upper bound on the distance between the recipient and `N(0)`. If the recipient
+is close to `N(0)`, this might not be ideal. In such cases, the recipient may add any number of
+dummy hops at the beginning of the blinded route by using `N(j) = N(r)`. The sender will not be
+able to distinguish those from normal blinded hops.
+
+Note that the recipient needs to fully validate each dummy hop to detect tampering.
+
+### Wallets and unannounced channels
+
+Route blinding is particularly useful for wallets that are connected to nodes via unannounced
+channels. Such wallets would use a single blinded hop, which effectively hides their `node_id`
+and `scid` from the sender. It obviously reveals to the blinded node that the next node is the
+final recipient, but a wallet that's not online all the time with a stable IP will never be able
```

Comment:
That's a real privacy problem, how to prevent LSPs to learn all your payment traffic ? Multi-homing LSP seems the best recommendation so far...

Reply:
Or simply using Tor and multiple `node_id`s. In that case a single LSP cannot distinguish that two of your `node_id`s belong to the same person.

Reply:
Okay but in that case you should avoid doing any kind of AMPs across channels spread on different `node_id`s because timing would be a leak? I will do a post on ml on all privacy-preserving tricks we may have for mobile-to-LSP

Reply:
> you should avoid doing any kind of AMPs across channels spread on different node_ids because timing would be a leak?

Yes definitely, when using multiple `node_id`s they should be used completely separately, as if they were from different users (so no AMP-ing between more than one of your `node_id`s).